File: packages/extractors/src/normalize.ts
/**
 * Split an identifier into normalized lowercase tokens — the raw material
 * for the ubiquitous-language concept layer.
 *   validateRefund   -> ["validate", "refund"]
 *   REFUND_WINDOW     -> ["refund", "window"]
 *   user/refund.flow  -> ["user", "refund", "flow"]
 */
export function splitIdentifier(id: string): string[] {
  return id
    .replace(/([a-z0-9])([A-Z])/g, "$1 $2") // camelCase boundary
    .replace(/[_\-./]+/g, " ") // snake / kebab / path separators
    .toLowerCase()
    .split(/\s+/)
    .filter((t) => t.length > 1 && !/^\d+$/.test(t));
}

export const normalize = (term: string): string =>
  splitIdentifier(term).join(" ");

File: extractors/src/main/kotlin/dev/stele/extractors/Normalize.kt
package dev.stele.extractors

private val CAMEL = Regex("([a-z0-9])([A-Z])")
private val SEPARATORS = Regex("[_\\-./:]+")
private val WHITESPACE = Regex("\\s+")
private val DIGITS = Regex("^\\d+$")

/**
 * Split an identifier into normalized lowercase tokens — raw material for the
 * ubiquitous-language concept layer.
 *   validateRefund  -> ["validate", "refund"]
 *   REFUND_WINDOW    -> ["refund", "window"]
 *   node:path/util   -> ["node", "path", "util"]
 */
fun splitIdentifier(id: String): List<String> =
    id.replace(CAMEL, "$1 $2")        // camelCase boundary
        .replace(SEPARATORS, " ")     // snake / kebab / path / module separators
        .lowercase()
        .split(WHITESPACE)
        .filter { it.length > 1 && !DIGITS.matches(it) }

fun normalize(term: String): String = splitIdentifier(term).joinToString(" ")
